tools/generate_db_schema.py

```diff
@@ -0,0 +1,227 @@
+                id_column_name = f"{cls.parent_table_class.table_name.rstrip('s')}_id"
+                table_name_to_id_column_name[cls.table_name] = id_column_name
+                schema += (
+                    f"  {id_column_name} {_id_type} "
+                    f"[pk, ref: - {cls.parent_table_class.table_name}.id]\n"
+                )
+        else:
+            table_name_to_id_column_name[cls.table_name] = "id"
+
+        for column in cls.columns:
+            schema += f"  {column.name} {_type_translation[column.ctype]}"
+            if column.name == "id":
+                schema += " [pk, increment]\n"
+            else:
+                schema += "\n"
+        schema += "}\n\n"
+
+    return schema, table_name_to_id_column_name
+
+
+def _generate_DBML_for_linker_tables(
+    linker_tables, primary_table_classes, table_name_to_id_column_name
+):
+    """Return DBML for `linker_tables`, using also the set of `primary_table_classes` and
+    the table_name -> id_column_name mapping
+
+    """
+    schema = ""
+    # Generate schema for linker tables
+    table_name_to_primary_class = {cls.table_name: cls for cls in primary_table_classes}
+    for parent_class, owned_object_list in linker_tables:
+        # Broken, so skip for now
+        if "component" in owned_object_list.joining_table_name:
+            continue
+
+        # Owner foreign key
+        schema += f"table {owned_object_list.joining_table_name}{{\n"
+
+        def link(cls_):
+            """Return column DBML for link class"""
+            # Determin the name of the column the foreign key points to
+            foreign_key_column_name = table_name_to_id_column_name[cls_.table_name]
+            if foreign_key_column_name == "id":
+                column_name = cls_.table_name.rstrip("s") + "_id"
+            else:
+                column_name = foreign_key_column_name
+
+            # Generate link text on the form:
+            #   calibration_id INTEGER [pk, ref: > ms_calibrations.calibration_id]
+            link_text = (
+                f"  {column_name} {_id_type} [pk, ref: > "
+                f"{cls_.table_name}.{foreign_key_column_name}]\n"
+            )
+            return link_text
+
+        # Generate columns for the two foreign keys in a linker class e.g:
+        #   calibration_id INTEGER [pk, ref: > ms_calibrations.calibration_id]
+        #   ms_cal_result_id INTEGER [pk, ref: > ms_cal_results.id]
+        owned_cls = table_name_to_primary_class[owned_object_list.owned_object_class]
+        for cls in (parent_class, owned_cls):
+            schema += link(cls)
+        schema += "}\n\n"
+
+    return schema
+
+
+def main():
+    """Generate all DB defining table information, convert to DBML and print and copy
+    to clipboard
+
+    """
+    primary_table_classes, linker_tables = collect_db_classes(verbose=False)
+    schema = generate_dbdiagramio_DBML(primary_table_classes, linker_tables)
+    pyperclip.copy(schema)
```

Comment:
Love it. Can also make a `ClipboardExporter` :) . We'll just need to at pyperclip to requirements.txt since it's not in the standard library

Reply:
Yeah, I considered this a development hack, so I didn't want to do anything with requirements for it and I don't remember if there was a difference in platform support etc. None, one or two of those could become like optional dependencies, with just nice error messages in the tool if they are missing.